// myfunc/myFunc.js
/**
 * Created by rbritton on 8/30/2016.
 */
'use strict';
process.env.NODE_ENV = 'test';

var configLoc = ("./_" + process.env.NODE_ENV + "_config");
var chai = require('chai');
var By = require('selenium-webdriver').By;
var webdriver = require('selenium-webdriver');
var console = require('console');
var fs = require('fs');
var moment = require('moment');
var config = require(configLoc);
var driver;
var expect = require('chai').expect;
var jwt = require('jsonwebtoken');
var until = require('selenium-webdriver').until;
var testPath;
var host = config.webURL;
var username = config.userID;
var password = config.oldPassword;

var createToken = function(app_id, app_key){
    let token = jwt.sign({
        app_id: app_id,
        scope: ['giver', 'taker'],
        scopes: ['giver', 'taker']
    }, app_key);
    console.log(token);
    return token;
};

var mkTestDir = function() {
    var dateStamp = Date.now();
    testPath = (config.testFilePath + dateStamp + "//");
    fs.mkdirSync(testPath);
    console.log(testPath);
    };
var screenPrint = function(fname){
    driver.takeScreenshot().then(
        function(image) {
            fs.writeFile(testPath + '\\' + fname, image, 'base64', function(err){
                console.log(err);
            });
        });
};
var findField = function(xpathVal){
driver.findElement(By.xpath(xpathVal));
};
var clearField = function(xpathVal){
    driver.findElement(By.xpath(xpathVal)).clear();
};
var setField = function(xpathVal, fieldVal){
    driver.findElement(By.xpath(xpathVal)).sendKeys(fieldVal);
};
var clickField = function(xpathVal){
    driver.findElement(By.xpath(xpathVal)).click();
};
var signOut = function(){
    clickField('//*[@href="/dashboard"]');
    driver.wait(function(){
        return driver.isElementPresent(webdriver.By.xpath('//*[text()="Dashboard"]'));
    }, 5000);
    findField('//*[text()="Dashboard"]');
    clickField('//*[@id="settings-option"]');
    driver.sleep(3000);
    clickField('//*[@class="sign-out svm-blue-text"]');
};
var waitToLoad = function(xpathVal){
    driver.wait(function(){
        return driver.isElementPresent(webdriver.By.xpath(xpathVal));
    }, 50000);
};
var gotoDashboard = function(){
    driver.get(host);
    waitToLoad('//*[text()="ServiceMaster Reviews"]');
    driver.sleep(3000);
    clickField('//*[@href="/dashboard"]');
    waitToLoad('//*[text()="Dashboard"]');
};
var verifyCompleted = function(name_param){
    gotoDashboard();
    clickField('//*[@href="/reports"]');
    waitToLoad('//*[text()="Reports"]');
    findField('//*[text()="Reports"]');
    driver.sleep(3000);
    clearField('//*[@class="reactive-table-input form-control"]');
    setField('//*[@class="reactive-table-input form-control"]',name_param);
    findField('//*[text()="completed"]');
    driver.sleep(3000);
};
var verifyPending = function(name_param){
    driver.sleep(3000);
    clickField('//*[@href="/reports"]');
    waitToLoad('//*[text()="Reports"]');
    findField('//*[text()="Reports"]');
    driver.sleep(3000);
    clickField('//*[@href="/dashboard"]');
    clickField('//*[@href="/reports"]');
    waitToLoad('//*[text()="Reports"]');
    findField('//*[text()="Reports"]');
    driver.sleep(3000);
    clearField('//*[@class="reactive-table-input form-control"]');
    setField('//*[@class="reactive-table-input form-control"]',name_param);
    findField('//*[text()="newly-created" or text()="processed" or text()="delivered"]');
    driver.sleep(3000);
};
var gotoProfile = function(){
    clickField('//*[@id="settings-option"]');
    clickField('//*[@href="/profile"]');
};
var createOrg = function(){
    clickField('//*[@id="organizations-option"]');
    clickField('//*[text()="Create Org"]');
    waitToLoad('//*[text()="Organization Information"]');
};
var verResp = function(res, status){
    expect(res).to.have.status(status)
};
var verFormat = function(res){
    //noinspection BadExpressionStatementJS
    expect(res).to.be.json
};
var verType = function(res, type){
    expect(res).to.be.a(type)
};
var verProp = function(field, name){
    expect(field).to.have.property(name)
};
var verVal = function(sentVal, setVal){
    expect(sentVal).to.equal(setVal)
};
var verValDeep = function(sentVal, setVal){
    expect(sentVal).to.eql(setVal)
};
var login = function(){
    driver.get(host);
    driver.wait(until.titleIs('Serv A Giver'), 2000);
    findField('//*[text()="ServiceMaster Reviews"]');
    clickField('//*[text() = "Login"]');
    waitToLoad('//*[text() = "Please Log In"]');
    setField('//*[@id = "email"]' ,username);
    setField('//*[@id = "password"]' ,password);
    clickField('//*[@name = "action"]');
    driver.sleep(3000);
};
exports.clearField = clearField;
exports.clickField = clickField;
exports.createOrg = createOrg;
exports.findField = findField ;
exports.gotoDashboard = gotoDashboard;
exports.gotoProfile = gotoProfile;
exports.setField = setField;
exports.signOut = signOut;
exports.verFormat = verFormat;
exports.verifyCompleted = verifyCompleted;
exports.verifyPending = verifyPending;
exports.verProp = verProp;
exports.verResp = verResp;
exports.verType = verType;
exports.verVal = verVal;
exports.verValDeep = verValDeep;
exports.waitToLoad = waitToLoad;
exports.mkTestDir = mkTestDir;
exports.screenPrint = screenPrint;
exports.createToken = createToken;
exports.login = login;
